### PR TITLE
Remove coin age priority and free transactions - implementation

### LINKED ISSUE #2673: minrelaytxfee doesn't accept 0
Using -minrelaytxfee or minrelaytxfee in bitcoin.conf results in:

Error: Invalid amount for -minrelaytxfee=<amount>: '0'

### LINKED ISSUE #6675: Remove TX priority and free transaction area from mempool, block creator
It was noted on #bitcoin-dev that many simplifications arise from removing the complex, dual-policy aspects of priority-based low fee transactions on policy estimation, block template creation and mempool code. This change updates Bitcoin Core to relay purely based on fee/kb rate, removing the concepts of priority and free transactions from the codebase.

fSendFreeTransactions, fLimitFree and similar controls have been operating in the field defaulting to fee-based choices already.

This change should make other changes such as @pstratem 's work to cap the mempool size dynamically much easier.

An example implementation exists (#6405) at https://github.com/jgarzik/bitcoin/tree/2015_remove_tx_priority
with the following notes/caveats:
- Status: A large change, lightly tested
- Transaction fee deltas are preserved - thus TX "priority" remains to some extent
- Arbitrary policies may be implemented via fee deltas
- prioritisetransaction RPC call continues to work, by adjusting the fee.
- needs focused review in the area of wallet, to ensure IsMine() transactions work fine through reorgs

### LINKED ISSUE #9601: Remove coin age priority and free transactions - discussion
I've opened #9602 to remove all coin age priority functionality from the codebase.

I think this has been discussed many times and hopefully it will not be the subject of another lengthy discussion, but just in case, I've opened this issue so any discussion not related to the technical merits of the implementation can happen here and the PR itself can remain free for discussion of the actual code.

This is targeted for 0.15, and shouldn't be merged until we ship 0.14, but it came up again recently and so I figured I might as well post the PR.  Also in the event that we feel anything needs to be pre-announced with the 0.14 release this will give us a chance to think about that, although I don't think there is anything pressing.

### PR DESCRIPTION
**Pull request by morcos**
Please see #9601 for all non-technical discussion.
Includes #9391 as the first commit
Please tag 0.15

This PR removes all coin age priority functionality from the codebase.

Some remaining open questions:
- The `prioritisetransaction` RPC API is broken.  Should we instead try to implement backwards compatible support, or is it fair to break it?  Previously it required exactly 3 arguments, now it requires exactly 2, so it will not result in unintended consequences.
- The 6th commit changes `sendrawtransaction` to no longer bypass minRelayTxFee by setting fLimitFree to false.  `prioritisetransaction` can always be used to achieve the same effect.  This seems an improvement in behavior, but isn't required (although the prioritise_transaction.py RPC test would need to be fixed if we don't include this).
- fLimitFree is now only ~~set~~ **false** for transactions we are trying to reaccept to the mempool from a disconnected block.  Should it also be used to bypass the mempool min fee in the event of mempool limiting.  I think yes, but can be left for a separate PR. **(EDIT for clarification:** currently if mempool limiting is active, even if fLimitFree is set false, transactions which don't pass the mempool min fee set by the last evicted transaction will still not be accepted to the mempool.  It is probably preferable that they are accepted and then the mempool will be limited again post acceptance and either there will have been room for them, they will be evicted, or they will be part of a more expensive package and something else will be evicted.**)**
-  The last commit is optional and introduces ability to set `-minrelaytxfee` to 0.  This isn't necessary, but seemed better to me especially now other ways to accept transactions without fee have been removed.

Fixes #9601
Fixes #6675
Fixes #2673 (thanks @MarcoFalke)

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK for removing it from the wallet. This will cause less headache when reviewing/reworking coin selection.

**Comment by MarcoFalke** [comment]
Needs rebase.

**Review by dcousens — APPROVED** [review]

**Comment by dcousens** [comment]
concept ACK,  light utACK

**Comment by btcdrak** [comment]
concept ACK.

**Comment by jtimon** [comment]
Concept ACK. Needs rebase.

File: src/init.cpp [hunk 1]
```diff
@@ -977,12 +977,12 @@ bool AppInitParameterInteraction()
...
     // if you set it to zero then
     // a transaction spammer can cheaply fill blocks using
-    // 1-satoshi-fee transactions. It should be set above the real
+    // 0-fee transactions. It should be set above the real
     // cost to you of processing a transaction.
     if (IsArgSet("-minrelaytxfee"))
     {
         CAmount n = 0;
-        if (!ParseMoney(GetArg("-minrelaytxfee", ""), n) || 0 == n)
+        if (!ParseMoney(GetArg("-minrelaytxfee", ""), n))
```
Comment (sdaftuar) [inline src/init.cpp]: nit: add the curly braces for this `if` while you're at it?

**Review by sdaftuar — COMMENTED** [review]
Nice!  Code review ACK, will test.

Left a few non-blocking nits for you to consider; I think they would be fine to address/discuss in future PRs if you preferred.

**Review by ryanofsky — COMMENTED** [review]
utACK a1ac40b49917638ebfa9d65a773a739519ea03da

- Changes make sense, though I think this might been have been easier to review if it consisted of a just few minimal commits that update the behavior, followed by a single larger commit that removes dead code without changing behavior.
- I didn't understand what you mean about potentially bypassing the mempool min fee in the event of mempool limiting. Could you say more about how this would work, what would be the advantages / disadvantages? Also, in the comment that "fLimitFree is now only set for transactions we are trying to reaccept to the mempool," assuming you meant "set false" not "set".
- Re: prioritisetransaction, one risk of reducing the number of arguments is that at some point in the future, somebody adds a new argument, and then old code calling the RPC with priority may appear to work but silently be broken. You could keep the priority_delta arg and require it to be null or 0 to prevent this. But I think getting rid of the argument is nice if you don't think it's likely someone will add a new prioritisetransaction argument in the future.

**Comment by sdaftuar** [comment]
ACK a1ac40b49917638ebfa9d65a773a739519ea03da

**Review by TheBlueMatt — COMMENTED** [review]
You may also wish to update the comment above paytxfee in contrib/debian/examples/bitcoin.conf

File: contrib/debian/examples/bitcoin.conf [hunk 2]
```diff
@@ -118,9 +118,6 @@
 
 # Transaction Fee Changes in 0.10.0
```
Comment (jnewbery) [inline contrib/debian/examples/bitcoin.conf]: I'd also lose the 'Transaction Fee Changes in 0.10.0' in the line above

File: src/init.cpp [hunk 3]
```diff
@@ -456,7 +456,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-printtoconsole", _("Send trace/debug info to console instead of debug.log file"));
     if (showDebug)
     {
-        strUsage += HelpMessageOpt("-printpriority", strprintf("Log transaction priority and fee per kB when mining blocks (default: %u)", DEFAULT_PRINTPRIORITY));
+        strUsage += HelpMessageOpt("-printpriority", strprintf("Log transaction fee per kB when mining blocks (default: %u)", DEFAULT_PRINTPRIORITY));
```
Comment (jnewbery) [inline src/init.cpp]: Agree with @sdaftuar . Commit at https://github.com/jnewbery/bitcoin/commit/b83daf977f38d98a3b3aa9ed5857bfd2f19b2a09 adds `-printfee` option. Feel free to pull into this PR if you want it, or I'll open another PR after this gets merged.

Comment (morcos) [inline src/init.cpp]: I always interpreted this as print the priority as in the sort order for mining.  I have no problem changing it, but I'll leave that for another PR if people want it.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3868,11 +3840,10 @@ bool CWallet::ParameterInteraction()
     }
     nTxConfirmTarget = GetArg("-txconfirmtarget", DEFAULT_TX_CONFIRM_TARGET);
     bSpendZeroConfChange = GetBoolArg("-spendzeroconfchange", DEFAULT_SPEND_ZEROCONF_CHANGE);
-    fSendFreeTransactions = GetBoolArg("-sendfreetransactions", DEFAULT_SEND_FREE_TRANSACTIONS);
     fWalletRbf = GetBoolArg("-walletrbf", DEFAULT_WALLET_RBF);
 
-    if (fSendFreeTransactions && GetArg("-limitfreerelay", DEFAULT_LIMITFREERELAY) <= 0)
-        return InitError("Creation of free transactions with their relay disabled is not supported.");
+    if (GetBoolArg("-sendfreetransactions", false))
+        InitWarning("The argument -sendfreetransactions is no longer supported.");
```
Comment (morcos) [inline src/wallet/wallet.cpp]: This was added in #9391 whose commit was included here, but then removed by the end of this PR.  Would have been cleaner to take it out of the first commit, but I'm going to aim for less history churn here, it doesn't hurt anything...

File: src/miner.cpp [hunk 5]
```diff
@@ -263,11 +263,7 @@ void BlockAssembler::AddToBlock(CTxMemPool::txiter iter)
 
     bool fPrintPriority = GetBoolArg("-printpriority", DEFAULT_PRINTPRIORITY);
     if (fPrintPriority) {
-        double dPriority = iter->GetPriority(nHeight);
-        CAmount dummy;
-        mempool.ApplyDeltas(iter->GetTx().GetHash(), dPriority, dummy);
-        LogPrintf("priority %.1f fee %s txid %s\n",
-                  dPriority,
+        LogPrintf("fee %s txid %s\n",
```
Comment (jnewbery) [inline src/miner.cpp]: meh. It's pretty obvious that this is feerate from context:

```
2017-03-03 14:19:33 fee 0.00005913 BTC/kB txid 3bbae9e9e8299f6335b5e99912cb6717df794212477a92fb425c2b5993d3bdc7
2017-03-03 14:19:33 fee 0.00004437 BTC/kB txid a3c976f58b52b8ab29eda3c434cb21d1d3ba1d30136b329a51ba4a3f493e6adf
```

File: src/txmempool.cpp [hunk 6]
```diff
@@ -930,18 +929,17 @@ void CTxMemPool::PrioritiseTransaction(const uint256& hash, double dPriorityDelt
             }
         }
     }
-    LogPrintf("PrioritiseTransaction: %s priority += %f, fee += %d\n", hash.ToString(), dPriorityDelta, FormatMoney(nFeeDelta));
+    LogPrintf("PrioritiseTransaction: %s fee += %d\n", hash.ToString(), FormatMoney(nFeeDelta));
```
Comment (morcos) [inline src/txmempool.cpp]: ok

File: src/net_processing.cpp [hunk 7]
```diff
@@ -1853,7 +1853,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                             LogPrint("mempool", "   invalid orphan tx %s\n", orphanHash.ToString());
                         }
                         // Has inputs but not accepted to mempool
-                        // Probably non-standard or insufficient fee/priority
+                        // Probably non-standard or insufficient fee
```
Comment (morcos) [inline src/net_processing.cpp]: will fix

File: src/miner.cpp (hunk 5 above)
Comment (morcos) [inline src/miner.cpp]: changed anyway

File: src/init.cpp (hunk 1 above)
Comment (morcos) [inline src/init.cpp]: ok

File: src/init.cpp [hunk 8]
```diff
@@ -977,12 +977,12 @@ bool AppInitParameterInteraction()
     // If you are mining, be careful setting this:
     // if you set it to zero then
     // a transaction spammer can cheaply fill blocks using
-    // 1-satoshi-fee transactions. It should be set above the real
+    // 0-fee transactions. It should be set above the real
```
Comment (jnewbery) [inline src/init.cpp]: Consider also changing the first sentence in this comment: `Fee-per-kilobyte amount considered the same as "free"` doesn't make much sense now that we don't relay free transactions by default.

Comment (morcos) [inline src/init.cpp]: ok

File: contrib/debian/examples/bitcoin.conf (hunk 2 above)
Comment (morcos) [inline contrib/debian/examples/bitcoin.conf]: ok

File: src/txmempool.cpp [hunk 9]
```diff
@@ -914,7 +914,7 @@ void CTxMemPool::PrioritiseTransaction(const uint256& hash, const CAmount& nFeeD
             }
         }
     }
-    LogPrintf("PrioritiseTransaction: %s fee += %d\n", hash.ToString(), FormatMoney(nFeeDelta));
+    LogPrintf("PrioritiseTransaction: %s feerate += %s\n", hash.ToString(), FormatMoney(nFeeDelta));
```
Comment (sdaftuar) [inline src/txmempool.cpp]: Good catch!

**Comment by sdaftuar** [comment]
re-ACK 7445fe9

**Comment by jnewbery** [comment]
I don't like the fact that prioritisetransaction is not back-compatible. Other than that, looks good. A few nits inline.

**Comment by morcos** [comment]
Nits addressed

Squashed 3b2b549 ([nopriority-tag2](https://github.com/morcos/bitcoin/commits/nopriority-tag2)) -> b421e6d

@jnewbery I didn't want to get into a whole big deal about what the current purpose of minrelaytxfee is so I left the comment minimally changed for now.

As for the backwards compatibility of prioritisetransaction, that was one of my open questions from the top of the PR.  I'll defer to group consensus, but my viewpoint is if we never change it then we're stuck with a stupid dummy field for all eternity, and if we are ever going to change it, now is the time to do so.  I think its a pretty safe change as it'll either work the old way or the new but they can't be mixed up.

**Comment by jnewbery** [comment]
> As for the backwards compatibility of prioritisetransaction...

My preference would be to maintain backwards compatibility, but I'll go along with group consensus and don't think this should delay merging this PR.

utACK.

**Comment by sdaftuar** [comment]
re-ACK b421e6ddcf00f220732f43742393452bb8bf4cdd

**Comment by TheBlueMatt** [comment]
utACK b421e6ddcf00f220732f43742393452bb8bf4cdd

**Comment by MarcoFalke** [comment]
utACK b421e6d

(only reviewed tests and wallet, did not look closely at miner and
mempool changes.)

Some notes about doc fixes for later commits:
* The change in check-doc.py seems no longer relevant
* The rpc breakage should be mentioned in the release notes
* The misleading comment in init.cpp about minrelaytxfee can be
removed alltogether. Better no doc than wrong doc and inline comments
are the wrong place for user documentation anyway.

**Comment by jtimon** [comment]
Note: this post will be updated as I progress in my review.
fast review ACK: https://github.com/bitcoin/bitcoin/pull/9602/commits/ddf58c75739efc7509d529002975ea447a269b3a
utACK: https://github.com/bitcoin/bitcoin/pull/9602/commits/12839cdd56ab5ebb06c59bbf9a064f4bd23b7398

File: src/rpc/mining.cpp [hunk 10]
```diff
@@ -258,31 +258,28 @@ UniValue getmininginfo(const JSONRPCRequest& request)
 // NOTE: Unlike wallet RPC (which use BTC values), mining RPCs follow GBT (BIP 22) in using satoshi amounts
 UniValue prioritisetransaction(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() != 3)
```
Comment (dcousens) [inline src/rpc/mining.cpp]: Agreed,  my only concern is this breaking API.   It will hurt someone.

**Review by dcousens — APPROVED** [review]
utACK b421e6d

**Pull request merged**